### PR TITLE
SRE-2118-homebrew-tinker-download-failing

### PR DESCRIPTION
**Pull request by paikattilrohin**
Closes https://snapsheettech.atlassian.net/browse/SRE-2118
## What
Homebrew upgraded to ruby 3 and we need to update our formula so that it works with the new version.

## Why
Jacob notice the following an error while updating tinker. 
```
jake@KQVPM7HPQT ~ % brew install snapsheet/core/tinker  
==> Fetching snapsheet/core/tinker
==> Fetching tinker gem from GitHub packages.
Downloaded tinker-1.0.1
==> Installing tinker from snapsheet/core
Error: An exception occurred within a child process:
  Psych::DisallowedClass: Tried to load unspecified class: Gem::Specification
jake@KQVPM7HPQT ~ % 
```

## How
Load the metadata using [Gem::Package library](https://docs.ruby-lang.org/en/master/Gem/Package.html). 

## Testing
Uninstall tinker using 
```
brew uninstall session-manager-plugin
brew untap snapsheet/core
brew uninstall tinker
```

checkout main and run 
```
brew install --formula --build-from-source Formula/tinker.rb
```

You should see an error.

```
Error: An exception occurred within a child process:
  Psych::DisallowedClass: Tried to load unspecified class: Gem::Specification
```

Checkout this branch and run 
```
brew reinstall --formula --build-from-source Formula/tinker.rb
```

You should see that you do not run into the error anymore.

### CODE REVIEW

**Comment by paikattilrohin** [comment]
Created another [bug](https://snapsheettech.atlassian.net/browse/SRE-2123) to track another issue @whatisdot and I found specific to ruby version 3.2.2(could be present in other ruby versions as well).

File: Formula/tinker.rb [hunk 1]
```diff
@@ -59,7 +59,7 @@ def ruby_version
   #
   # @return [Object] Reference to metadata object loaded from a gem.
   def metadata
-    @metadata ||= YAML.load(`tar -xOf #{tinker_gem_path} metadata.gz | gzip -dc`)
+    @metadata ||= YAML.load(`tar -xOf #{tinker_gem_path} metadata.gz | gzip -dc`, permitted_classes: [Gem::Specification, Gem::Version, Gem::Dependency, Gem::Requirement, Gem::Platform, Time, Symbol])
```
Comment (paikattilrohin) [inline Formula/tinker.rb:62]: These were used as a reference for this solution.  
[Upgrading to Ruby 3.1 causes Psych::DisallowedClass exception when using YAML.load_file](https://stackoverflow.com/questions/71332602/upgrading-to-ruby-3-1-causes-psychdisallowedclass-exception-when-using-yaml-lo) 
https://www.rubydoc.info/stdlib/psych/Psych.safe_load
Screen Shot 2024-02-23 at 9 58 20 AM

File: Formula/tinker.rb [hunk 2]
```diff
@@ -59,7 +59,7 @@ def ruby_version
   #
   # @return [Object] Reference to metadata object loaded from a gem.
   def metadata
-    @metadata ||= YAML.load(`tar -xOf #{tinker_gem_path} metadata.gz | gzip -dc`)
+    @metadata ||= YAML.load(`tar -xOf #{tinker_gem_path} metadata.gz | gzip -dc`, permitted_classes: [Gem::Specification, Gem::Version, Gem::Dependency, Gem::Requirement, Time, Symbol])
```
Comment (whatisdot) [inline Formula/tinker.rb:62]: I was able to get the metadata by using this standard library class. 
```ruby
require 'rubygems/package'

@metadata ||= Gem::Package.new(tinker_gem_path).spec
```

Comment (paikattilrohin) [reply Formula/tinker.rb:62]: Nice! It worked for me also.

**Review by whatisdot — APPROVED** [review]
Tested and works as expected. This solution is a lot cleaner 👍 

Can you update the PR description to reflect this change in approach, and include a link to the native library doc page?

**Review by d10zero — APPROVED** [review]
CR: LG
FR: confirmed tinker download thru brew works properly with this up

**Pull request merged**